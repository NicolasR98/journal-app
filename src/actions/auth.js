import { createUserWithEmailAndPassword, getAuth, signInWithPopup, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from '../types/types'

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

// To handle an async function, the helper must return another function
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {
        setTimeout(() => {
            dispatch(login(123, 'nicolas'));
        }, 3500);
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // Get auth and make a petition to firebase to create an user
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, password)
            // Once created, make a post req to firebase to update the current
            // user displayName (async)
            .then(async ({ user }) => {
                await updateProfile(user, { displayName: name });
                dispatch(
                    login(user?.uid, user.displayName)
                );
            })
            .catch(err => err);
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({ user }) => {
                dispatch(
                    login(user?.uid, user?.displayName)
                );
            })
            .catch(e => console.log(e));
    };
};