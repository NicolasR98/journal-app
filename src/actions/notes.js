import { addDoc, collection, deleteDoc, doc, updateDoc } from "firebase/firestore";
import { db } from "../firebase/firebaseConfig";

import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";

import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime(),
        };

        const { uid } = getState().auth;
        const docRef = await addDoc(collection(db, `${uid}`, 'journal/notes'), newNote);
        dispatch(activeNote(docRef?.id, newNote));
        dispatch(addNewNote(docRef.id, newNote));
    };
};

const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note,
    },
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note,
    },
});

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes,
});

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        if (!note.url) {
            delete note.url;
        };

        const noteToFirestore = { ...note };
        delete noteToFirestore.id;

        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);

        await updateDoc(noteRef, noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));

        Swal.fire('Note Saved!', noteToFirestore.title, 'success')
    };
};

export const refreshNote = (id, note) => ({
    type: types.notesUpdate,
    payload: {
        id,
        note: {
            id,
            ...note,
        },
    }
});

export const startUpload = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;

        Swal.fire({
            title: 'Uploading file...',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => { Swal.showLoading() }
        })

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        dispatch(startSaveNote(activeNote));

        Swal.close();
    };
};

export const startDelete = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;

        const noteRef = doc(db, `${uid}/journal/notes/${id}`);

        // Delete note on firestore
        await deleteDoc(noteRef);

        // Delete note on store
        dispatch(deleteNote(id));

        Swal.fire('Note deleted!', 'Note deleted successfully', 'success')
    };
};

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id,
});

export const logoutCleanNotes = () => ({
    type: types.notesLogoutClean,
});