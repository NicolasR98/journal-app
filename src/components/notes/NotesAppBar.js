import React from 'react';
import { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote, startUpload } from '../../actions/notes';

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const { active } = useSelector(state => state.notes);

  const inputFileSelector = useRef();

  const handleSave = () => {
    dispatch(startSaveNote(active));
  };

  const handlePictureUpload = () => {
    inputFileSelector.current.click();
  };

  const handleFileChange = ({ target }) => {
    const [file] = target.files;
    if (file) {
      dispatch(startUpload(file))
    }

  }

  return (
    <div className='notes__appbar'>
      <span>30th january of 2022</span>

      <input
        ref={inputFileSelector}
        type='file'
        name='file'
        style={{ display: 'none' }}
        onChange={handleFileChange}
      />

      <div>
        <button
          onClick={handlePictureUpload}
          className='btn'
        >
          Upload picture
        </button>

        <button
          className='btn'
          onClick={handleSave}
        >
          Save
        </button>
      </div>
    </div>
  );
};
