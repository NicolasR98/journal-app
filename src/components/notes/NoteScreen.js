import React from 'react';
import { useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
    const { active: note } = useSelector(state => state.notes);
    const [formValues, handleInputChange] = useForm(note);

    const { body, title } = formValues;

    return (
        <div className='notes__main-content'>
            <NotesAppBar />

            <div className='notes__content'>
                <input
                    type='text'
                    placeholder='Some awesome title'
                    className='notes__title-input'
                    value={title}
                    onChange={handleInputChange}
                />

                <textarea
                    placeholder='What did you do today?'
                    className='notes__textarea'
                    value={body}
                    onChange={handleInputChange}
                />

                {(note.url) &&
                    <div className='notes__image'>
                        <img
                            src='https://images.unsplash.com/photo-1520962922320-2038eebab146?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80'
                            alt='Roadtrip'
                        />
                    </div>
                }
            </div>
        </div>
    );
};
