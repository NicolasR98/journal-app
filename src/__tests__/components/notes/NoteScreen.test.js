import React from 'react';
import { mount } from "enzyme";
import { Provider } from "react-redux";

import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';

import { NoteScreen } from "../../../components/notes/NoteScreen";
import { activeNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn(),
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const uid = 'TESTING_UID';
const initialState = {
    auth: {
        uid,
    },
    notes: {
        active: {
            id: '1234',
            body: 'Testing body',
            title: 'Testing title',
            date: 0,
        },
        notes: [],
    },
};

let store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
        <NoteScreen />
    </Provider>
)

describe('Tests on <NoteScreen />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('should display correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should trigger the correct action when the input changes', () => {
        const title = 'Title updated';
        const expected = {
            id: '1234',
            body: 'Testing body',
            title,
            date: 0,
        };
        const inputTitle = wrapper.find('input[name="title"]');
        inputTitle.simulate('change', {
            target: {
                name: 'title',
                value: title
            },
        });

        expect(activeNote).toHaveBeenLastCalledWith(expected.id, expected);
    });
});