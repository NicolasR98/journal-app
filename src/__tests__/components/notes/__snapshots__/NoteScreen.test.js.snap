// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <NoteScreen /> should display correctly 1`] = `
<div
  className="notes__main-content animate__animated animate__fadeIn animate__faster"
>
  <div
    className="notes__appbar"
  >
    <span>
      30th january of 2022
    </span>
    <input
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Upload picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Testing title"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What did you do today?"
      value="Testing body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
